servant/src/Servant/API/Authentication.hs

```diff
@@ -0,0 +1,41 @@
+{-# LANGUAGE DataKinds          #-}
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE KindSignatures     #-}
+{-# LANGUAGE PolyKinds          #-}
+{-# LANGUAGE TypeFamilies       #-}
+{-# OPTIONS_HADDOCK not-home    #-}
+module Servant.API.Authentication
+( AuthPolicy (..)
+, AuthProtect (..)
+, AuthProtected (..)
+, BasicAuth (..)
+, JWTAuth (..)
+) where
+
+
+import           Data.ByteString (ByteString)
+import           Data.Typeable   (Typeable)
+import           GHC.TypeLits    (Symbol)
+import           Data.Text       (Text)
+
+-- | we can be either Strict or Lax.
+-- Strict: all handlers under 'AuthProtect' take a 'usr' argument.
+--         when auth fails, we call user-supplied handlers to respond.
+-- Lax: all handlers under 'AuthProtect' take a 'Maybe usr' argument.
+--      when auth fails, we call the handlers with 'Nothing'.
+data AuthPolicy = Strict | Lax
+
+-- | the combinator to be used in API types
+data AuthProtect authdata usr (policy :: AuthPolicy)
+
+-- | what we'll ask user to provide at the server-level when we see a
+-- 'AuthProtect' combinator in an API type
+data family AuthProtected authdata usr subserver :: AuthPolicy -> *
```

Comment:
(Continuing on with `->` ~ `(,)`:)
Some of the interesting possibilities that come up when the server is regular are e.g. transforming things into a `Reader` monad over e.g. a header (in the case of `->`); for `(,)`, we have the `State` monad, which can be similarly mechanically transformed. Off the top of my head, this seems promising: we can imagine writing subapis with nested but different auth requirements in a state monad. Pretty fancy - not sure if practically relevant in this specific case, but assuming `(,)` cases multiply, something is bound to come up.

This brings up the question of order wrt `,`. From the `State` perspective, it seems like we'd want the subserver on the LHS. OTOH from the perspective of the tuple functor instance, we'd want it on the RHS...